cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java

```diff
@@ -923,6 +928,142 @@ public StopProposalResult stopProposalExecution(StopProposalParameters parameter
+                                       + "its replication factor.", partitionInfo.topic(), partitionInfo.partition(), offlineReplicas));
+          }
+          List<Integer> currentAssignedReplica = new ArrayList<>(partitionInfo.replicas().length);
+          List<Integer> newAssignedReplica = new ArrayList<>();
+          Set<String> currentOccupiedRack = new HashSet<>();
+          // Make sure the current replicas are in new replica list.
+          for (Node node : partitionInfo.replicas()) {
+            currentAssignedReplica.add(node.id());
+            newAssignedReplica.add(node.id());
+            currentOccupiedRack.add(rackByBroker.get(node.id()));
+          }
+          // Add new replica to partition in rack-aware(if possible), round-robin way.
+          while (newAssignedReplica.size() < replicationFactor) {
+            if (!currentOccupiedRack.contains(racks.get(rackCursor)) || currentOccupiedRack.size() == racks.size()) {
+              String rack = racks.get(rackCursor);
+              int cursor = cursors[rackCursor];
+              newAssignedReplica.add(brokersByRack.get(rack).get(cursor));
+              currentOccupiedRack.add(rack);
+              cursors[rackCursor] = (cursor + 1) % brokersByRack.get(rack).size();
+            }
+            rackCursor = (rackCursor + 1) % racks.size();
+          }
+          proposals.add(new ExecutionProposal(new TopicPartition(topic, partitionInfo.partition()),
+                                              0,
+                                              partitionInfo.leader().id(),
+                                              currentAssignedReplica,
+                                              newAssignedReplica));
+        } else if (partitionInfo.replicas().length > replicationFactor) {
+          Set<Integer> offlineReplicas = offlineReplicasForPartition(partitionInfo, cluster.nodes());
+          if (!offlineReplicas.isEmpty()) {
+            throw new RuntimeException(String.format("Topic partition %s-%d has offline replicas on brokers %s, unable to update "
+                                       + "its replication factor.", partitionInfo.topic(), partitionInfo.partition(), offlineReplicas));
+          }
+          List<Integer> currentAssignedReplica = new ArrayList<>(partitionInfo.replicas().length);
+          List<Integer> newAssignedReplica = new ArrayList<>();
+          // Make sure the leader replica is in new replica list.
+          newAssignedReplica.add(partitionInfo.leader().id());
+          for (Node node : partitionInfo.replicas()) {
+            currentAssignedReplica.add(node.id());
+            if (newAssignedReplica.size() < replicationFactor && node.id() != newAssignedReplica.get(0)) {
+              newAssignedReplica.add(node.id());
+            }
+          }
+          proposals.add(new ExecutionProposal(new TopicPartition(topic, partitionInfo.partition()),
+                                              0,
```

Comment:
Can we move `0` to a static variable with a `TODO` message that points to the relevant Github issue for future patch?